### PR TITLE
workflow lineage

### PR DESCRIPTION
**Pull request by LiemLin**
## *Tips*
- *Thanks very much for contributing to Apache DolphinScheduler.*
- *Please review https://dolphinscheduler.apache.org/en-us/community/index.html before opening a pull request.*

## What is the purpose of the pull request

This pull request adds workflow lineage

## Brief change log

This change added tests and can be verified as follows:

  - *Added dolphinscheduler-api tests for unit test*
  - *Manually verified the change by testing locally.*

### CODE REVIEW

**Comment by dailidong** [comment]
great job,  haha

please solve Conflicting files locally and add apache license on the new file header, thx

**Comment by dailidong** [comment]
I see SonarCloud Quality Gate failed.  please check !

**Comment by LiemLin** [comment]
> I see SonarCloud Quality Gate failed. please check !

Ok, I will see why local works but SonarCloud doesn't

**Comment by lgcareer** [comment]
I pull your code and test,I use the postgresql db and occur one error.
image

**Comment by LiemLin** [comment]
> I pull your code and test,I use the postgresql db and occur one error.
> image

json_extract is a function of mysql, I'm going to have a generic implementation

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkFlowLineageService.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        if(ids == null || ids.size() == 0){
+            for(WorkFlowLineage workFlowLineage:workFlowLineageList) {
+                idsV.add(workFlowLineage.getWorkFlowId());
+            }
+        }
+        List<WorkFlowRelation> workFlowRelations = new ArrayList<>();
+        getWorkFlowRelationRecursion(idsV, workFlowRelations);
+
+        Set<Integer> idSet = new HashSet<>();
+        //如果传入参数不为空，则需要补充下游工作流明细属性
```
Comment (lgcareer) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkFlowLineageService.java:75]: You can translate this to english.

Comment (LiemLin) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkFlowLineageService.java:75]: OK, thanks for your remind

**Comment by lgcareer** [comment]
Thanks!! I changed the db,used mysql db,it can work.I made a cyclic dependency then it occured StackOverflowError.
Such as the follow steps.
1.process definiton A depend on process definition B.
2.process definiton B depend on process definition C.
3.process definiton C depend on process definition A.
Then it will occur StackOverflowError.
image

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkFlowLineageService.java (hunk 1 above)
Comment (lgcareer) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkFlowLineageService.java:75]: You're welcome.I saw you are good at both frontend and backend.It's great!

**Comment by LiemLin** [comment]
> Thanks!! I changed the db,used mysql db,it can work.I made a cyclic dependency then it occured StackOverflowError.
> Such as the follow steps.
> 1.process definiton A depend on process definition B.
> 2.process definiton B depend on process definition C.
> 3.process definiton C depend on process definition A.
> Then it will occur StackOverflowError.
> image

Endless loop。。。 Isn't workflow lineage a directed acyclic graph, can it be cycled？

**Comment by LiemLin** [comment]
> > Thanks!! I changed the db,used mysql db,it can work.I made a cyclic dependency then it occured StackOverflowError.
> > Such as the follow steps.
> > 1.process definiton A depend on process definition B.
> > 2.process definiton B depend on process definition C.
> > 3.process definiton C depend on process definition A.
> > Then it will occur StackOverflowError.
> > image
> 
> Endless loop。。。 Isn't workflow lineage a directed acyclic graph, can it be cycled？
> I will see how to prevent an endless loop
image

Like this? Is this cycle reasonable?

**Comment by lgcareer** [comment]
When add a depend node,it didn't vefiy whether will make a cylic dependency.

image
Just like the above image

**Comment by LiemLin** [comment]
> > I pull your code and test,I use the postgresql db and occur one error.
> > image
> 
> json_extract is a function of mysql, I'm going to have a generic implementation

hi , the postgresql support has been increased

**Comment by dailidong** [comment]
good job, there still some problems need to be disscussed , you can add my mail(lidongdai@apache.org) or wechat(510570367) , when added , please tell my your github id, thx

**Comment by lgcareer** [comment]
> > > I pull your code and test,I use the postgresql db and occur one error.
> > > image
> > 
> > 
> > json_extract is a function of mysql, I'm going to have a generic implementation
> 
> hi , the postgresql support has been increased

Great.Did you fix the  cylic dependency problem that result in  StackOverflowError?

**Comment by LiemLin** [comment]
> > > > I pull your code and test,I use the postgresql db and occur one error.
> > > > image
> > > 
> > > 
> > > json_extract is a function of mysql, I'm going to have a generic implementation
> > 
> > 
> > hi , the postgresql support has been increased
> 
> Great.Did you fix the cylic dependency problem that result in StackOverflowError?

Yes, I tested the cylic dependency locally and can work

**Comment by LiemLin** [comment]
> good job, there still some problems need to be disscussed , you can add my mail([lidongdai@apache.org](mailto:lidongdai@apache.org)) or wechat(510570367) , when added , please tell my your github id, thx

Ok， has been added

**Comment by dailidong** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=VULNERABILITY) (and [Security Hotspot](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=SECURITY_HOTSPOT) to review)
> [Code Smell](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=2421&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=2421&metric=new_coverage&view=list) [12.1% Coverage](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=2421&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=2421&metric=new_duplicated_lines_density&view=list) [6.4% Duplication](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=2421&metric=new_duplicated_lines_density&view=list)

for now UT class should register to pom.xml , like this:
image

**Comment by qiaozhanwei** [comment]
@lgcareer please trace this PR . Thx

**Comment by lgcareer** [comment]
@LiemLin Hi,I tested and now the cylic dependency didn't error.It looks like it didn't show the relation if it has a cylic dependency.

**Comment by break60** [comment]
Do not change the'.env' file

**Comment by break60** [comment]
It is not for you to delete the'.env', it is for you to restore to the original look

**Pull request merged**